File: dsa/feb1.2024/InsertionSort.java
import java.util.Arrays;

public class InsertionSort {
    public static void main(String[] args) {
        int[] nums = {5, 4, 3, 4, 3, 2, 4, 1, 2, 3, 4, 3, 100};
        for (int i = 1; i < nums.length; i++) {
            int j = i;
            while (j > 0 && nums[j - 1] > nums[j]) {
                int temp = nums[j];
                nums[j] = nums[j - 1];
                nums[j - 1] = temp;
                j -= 1;
            }
        }
        System.out.println(Arrays.toString(nums));
    }
}

File: dsa/feb1.2024/insertion_sort.py
nums = [10, 2, 6, 5, 1, 3, 4]

def insertion_sort(nums):
    for i in range(1, len(nums)):
        # j = i-1
        # while nums[j] > nums[j+1] and j >= 0:
        #     nums[j], nums[j+1] = nums[j+1], nums[j]
        #     j -= 1
        for j in range(i - 1, -1, -1):
            if (nums[j] > nums[j+1]):
                nums[j], nums[j+1] = nums[j+1], nums[j]
            else:
                break 
    return nums

insertion_sort(nums)
print(nums)